### PR TITLE
Bug 1870289: LOG-755: Make ClusterLogForwarder cluster scoped

### PR DESCRIPTION
**Pull request by alanconway**

### CODE REVIEW

**Comment by ewolinetz** [comment]
are there other implications from this? does it change the operator RBAC at all?

**Comment by alanconway** [comment]
> are there other implications from this? does it change the operator RBAC at all?

Not sure, are there any additional tests we should run to make sure it's OK?

**Comment by alanconway** [comment]
/retest
apparently the "v2" error is a CI platform flake.

**Comment by periklis** [comment]
@alanconway @ewolinetz I don't see a showstopper regarding RBAC. However, passing namespaces to the manager in `main.go` could be: https://sdk.operatorframework.io/docs/crds-scope/#crd-cluster-scoped-usage --> Is this an issue?

**Comment by jcantrill** [comment]
@alanconway I rebased in a separate commit to see if tests would pass but please collapse this change into a single commit

**Comment by alanconway** [comment]
> @alanconway @ewolinetz I don't see a showstopper regarding RBAC. However, passing namespaces to the manager in `main.go` could be: https://sdk.operatorframework.io/docs/crds-scope/#crd-cluster-scoped-usage --> Is this an issue?

That URL has been broken, I found this one: https://sdk.operatorframework.io/docs/legacy-common/operator-scope/
Does it cover the same issue?

**Comment by periklis** [comment]
> > @alanconway @ewolinetz I don't see a showstopper regarding RBAC. However, passing namespaces to the manager in `main.go` could be: https://sdk.operatorframework.io/docs/crds-scope/#crd-cluster-scoped-usage --> Is this an issue?
> 
> @periklis That URL has been broken, I found this one: https://sdk.operatorframework.io/docs/legacy-common/operator-scope/
> Does it cover the same issue?

Yes, in detail: https://sdk.operatorframework.io/docs/legacy-common/operator-scope/#changes-required-for-a-cluster-scoped-operator

File: docs/HACKING.md [hunk 1]
```diff
@@ -160,3 +159,19 @@ will attempt to clone them to a temporary directory.
...
+```
+make deploy
+...
+waiting the deployment openshift-logging/cluster-logging-operator - remaining attempts: 0
+```
+
+* `oc project openshift-logging` following commands assume this. 
+* `oc get all` missing/unexpected resources?
+* `oc describe all` check conditions and events.
+
```
Comment (periklis) [inline docs/HACKING.md:177]: TBH, I came up with a more concrete approach to debug OLM deployments based on the chain of events and CRs created and reconciled by the two OLM operators namely the `catalog-operator` and the `olm-operator`. The resources collected below are part of our `logging.dump.sh` [script](https://github.com/openshift/origin-aggregated-logging/blob/master/hack/logging-dump.sh).

In general the OLM operator installation process works like this (explained more visually in our Openshift docs though):
1. Operator installation starts by creating a **subscription** for the registry's catalog. First hint to check for broken requirements is : `oc get subscription -o yaml`
2. The **subscription** is translated in an **install plan** that tells OLM what resources to install, i.e. CSV, RBAC, CRDs, ServiceAccount. So the next hint is: `oc get installplan -o yaml`
3. If the **install plan** was applied correctly our should see the installed resources, ie. **CSV, rbac, CRDs, ServiceAccount, Deployment** in your target namespace: So next hint is: `oc get csv -o yaml`
4. If the pods don't start, you should check the logs of your operator, e.g `oc logs cluster-logging-operator-...`
5. Finally, checking the logs of the operator-framework can also give hints on what is broken:
5.1 Logs for **catalog-operator**: `oc -n openshift-operator-lifecycle-manager logs catalog-operator-...`
5.2 Logs for **olm-operator**: `oc -n openshift-operator-lifecycle-manager logs olm-operator-...`

File: manifests/4.6/cluster-logging.v4.6.0.clusterserviceversion.yaml [hunk 2]
```diff
@@ -310,9 +319,7 @@ spec:
                 - cluster-logging-operator
                 env:
                   - name: WATCH_NAMESPACE
-                    valueFrom:
-                      fieldRef:
-                        fieldPath: metadata.annotations['olm.targetNamespaces']
+                    value: ""
```
Comment (periklis) [inline manifests/4.6/cluster-logging.v4.6.0.clusterserviceversion.yaml:322]: Although this is the recommendation from the operator-sdk docs, I believe it now allows to watch also for `ClusterLogging` resources in all namespaces. WDYT?

IMHO we can fix the limitation in the `clusterlogging_controller` by extending the watch predicate to watch for create/update/delete/generic events where the event object is in `openshift-logging`. (See https://github.com/openshift/cluster-logging-operator/blob/master/pkg/controller/clusterlogging/clusterlogging_controller.go#L41)

Comment (ewolinetz) [reply manifests/4.6/cluster-logging.v4.6.0.clusterserviceversion.yaml:322]: I'm not sure we need to change this... `ElasticsearchOperator` also watches all namespaces but follows this pattern as well:
https://github.com/openshift/elasticsearch-operator/blob/master/manifests/4.6/elasticsearch-operator.v4.6.0.clusterserviceversion.yaml#L269-L272

Comment (periklis) [reply manifests/4.6/cluster-logging.v4.6.0.clusterserviceversion.yaml:322]: @ewolinetz I am not sure if this compatible with the updated operator-sdk 0.18 and the underlying controller-runtime 0.6.x

Comment (alanconway) [reply manifests/4.6/cluster-logging.v4.6.0.clusterserviceversion.yaml]: I reverted this - looks like it not necessary (for now, in future if we implement per-namespace forwarders we do that) Need another LGTM please.

**Comment by ewolinetz** [comment]
before I give this a `/lgtm` e2e tests need to be addressed (they're failing as a result of changes in this PR)

File: HACKING.md [hunk 3]
```diff
@@ -1,162 +1,193 @@
...
+   - *Deployed tests:* Build and deploy CLO image to your own cluster, allows you to run tests selectively.
+   - *Local process:* requires a cluster with dependencies deployed (elasticsearch), but runs CLO as a *local process*. Faster and easier to debug.
+
+All PRs should pass unit tests. CI runs all tests automatically and won't merge a PR till they pass, so it is OK to create a PR and let CI run the e2e tests if you've done enough testing to be confident. This allows for human review and feedback while CI tests are running.
+
+To get a code review of unfinished work, create a PR with "[WIP]" at the start of the description. The CI system will not merge it until the "[WIP]" is removed and all tests are passing.
+
+## Setting up a test cluster
+
+Should work with any k8s cluster; tested with AWS and [Code Ready Containers](https://developers.redhat.com/products/codeready-containers/download). See below for extra steps with some specific cluster providers.
```
Comment (jcantrill) [inline HACKING.md:40]: This is not correct given we have dependencies on Openshift Auth apis still in Elasticsearch.  I@ewolinetz we have not removed those correct?

File: HACKING.md [hunk 4]
```diff
@@ -1,162 +1,193 @@
...
+oc registry login --to my-secret-file
+oc logout
+# This creates or adds to my-secret-file
+```
+
+Once created, you can `export KUBECONFIG=/path/to/my/cluster/config` or copy your config to `$HOME/.kube/config`.
+
+**Note**: Don't symlink `$HOME/.kube/config`. Some commands may overwrite this file and break the symlink, causing much confusion.
+
+Log in to your cluster as a user with the `cluster-admin` role. User `kubeadmin` is usually predefined with this role.
```
Comment (jcantrill) [inline HACKING.md:58]: My experience has been this is not correct.  You have to explicitly grant that role which seems strange.  You can confirm this when logging into the kibana interface and it does not think you are an admin

File: HACKING.md [hunk 5]
```diff
@@ -1,162 +1,193 @@
...
+To build, deploy and test the CLO image in your own cluster:
+* `make deploy`: Build CLO image and deploy to cluster with elasticsearch.
+* `make undeploy`: Undo `make deploy`.
+
+To run CLO as a local process:
+* `make run`: Run CLO as a local process. Does not require `make deploy`.
+* `make debug`: Run CLO in the `dlv` debugger.
+
+After running `make deploy`, `make run` or `make debug`, you can run e2e tests:
+* `make test-cluster` Run all e2e tests against the running CLO.
```
Comment (jcantrill) [inline HACKING.md:90]: i like `test-cluster` but it  doesn't seem that it obviously describes e2e or the path where you will find the tests

File: README.md [hunk 6]
```diff
@@ -26,7 +26,7 @@ The CLO (Cluster Logging Operator) *owns* the following APIs:
 
 To install a released version of cluster logging see the [Openshift Documentation](https://docs.openshift.com/)
 
-To experiment or contribute to the development of cluster logging, see [HACKING.md](./docs/HACKING.md)
+To experiment or contribute to the development of cluster logging, see [HACKING.md](./HACKING.md)
```
Comment (jcantrill) [inline README.md:29]: Is there a need to move the doc?  more discoverable?

Comment (alanconway) [reply README.md:29]: I think it's more conventional - its the pattern in other openshift repos.

Comment (alanconway) [reply README.md:29]: I've seen it in other places too - README for users HACKING for contributors.

**Comment by alanconway** [comment]
@jcantrill I removed all the README/HACKING stuff, can you LGTM the cluster scope change?
I'll make a new PR for the other bits when I've reviewed your comments.

**Comment by alanconway** [comment]
/hold
retest spinning

File: HACKING.md (hunk 3 above)
Comment (alanconway) [reply HACKING.md]: All this stuff taken out of this PR, will do a separate PR to improve the HACKING doc.

File: HACKING.md (hunk 4 above)
Comment (alanconway) [reply HACKING.md]: All this stuff taken out of this PR, will do a separate PR to improve the HACKING doc.

File: README.md (hunk 6 above)
Comment (alanconway) [reply README.md]: All this stuff taken out of this PR, will do a separate PR to improve the HACKING doc.

File: HACKING.md (hunk 5 above)
Comment (alanconway) [reply HACKING.md]: Removed, this function is now done by test-e2e-local.

File: docs/HACKING.md (hunk 1 above)
Comment (alanconway) [reply docs/HACKING.md]: All this stuff taken out of this PR, will do a separate PR to improve the HACKING doc.

**Pull request merged**